### PR TITLE
Update Text.locate, Text.locate_all, Text.index_of and Text.last_index_of #5119

### PR DESCRIPTION
**Pull request by GregoryTravis**
### Pull Request Description

Remove regex support from .locate and .locate_all; regex functionality is moved to .match and .match_all where appropriate. This is in preparation for simplifying regex support across the board.

Also change Matching_Mode types to a single type with two variants.

Note: the matcher parameter to .locate and .locate_all has been replaced by a case_sensitivity parameter, of type Case_Sensitivity, which differs in that it also has a Default option. Default is treated as Sensitive.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -305,6 +305,7 @@
   simpler API.][4078]
 - [Updated `Table.set` to new API. New `Column.parse` function and added case
   sensitivity to `Filter_Condition` and column functions.][4097]
+- [Moved regex functionality out of `Text.locate` and `Text.locate_all` into`Text.match` and `Text.match_all`.][5679]
```
Comment (jdunkerley) [inline CHANGELOG.md]:
```suggestion
- [Moved regex functionality out of `Text.locate` and `Text.locate_all` into
  `Text.match` and `Text.match_all`.][5679]
```

Wrapping at 80 chars.

**Review by jdunkerley — APPROVED** [review]
From a quick review looks good. @radeusgd please review as well.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 2]
```diff
@@ -229,11 +229,41 @@ Text.characters self =
...
+     matching mode where the matches are searched for from the back.
+
+   > Example
+     Comparing Matching in Last Mode in Regex and Text mode
+
+         "aaa".match "aa" mode=Matching_Mode.Last matcher=Text_Matcher == "aaa"
+         "aaa".match "aa" mode=Matching_Mode.Last matcher=Regex_Matcher == "aaa"
+
+         "aaa aaa".match "aa" mode=Matching_Mode.Last matcher=Text_Matcher == "aaa aaa"
+         "aaa aaa".match "aa" mode=Matching_Mode.Last matcher=Regex_Matcher == "aaa aaa"
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Once we stripped off the locations, this examples does not really show any difference - the return values are exactly the same. I think if we still want to illustrate the difference we'd need to do:
```suggestion
   > Example
     Comparing Matching in Last Mode in Regex and Text mode

         "aAa".match "aa" mode=Matching_Mode.Last matcher=Text_Matcher.Case_Insensitive == "Aa"
         "aAa".match "aa" mode=Matching_Mode.Last matcher=(Regex_Matcher.Value case_sensitivity=Case_Sensitivity.Insensitive) == "aA"
```

btw. the example seem to have been out of date, so they need updating anyway. (They were already out of date before your changes, seems that we forgot to update them when reworking the matchers)

(I did not run the examples above, please check if they return what is expected, in fact - please just add them to the tests, replacing `==` with ` . should_equal`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Regex_Mode.enso [hunk 3]
```diff
@@ -4,7 +4,7 @@
    to matching on the `Full` content of the input text.
 
 import project.Data.Numbers.Integer
-import project.Data.Text.Matching_Mode
+import project.Data.Text.Matching_Mode.Matching_Mode
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Regex_Mode.enso:7]: I cannot add a comment below but since we have merged `First` and `Last` to a common supertype, the line below must be updated accordingly:

```diff
- Bounded (start : Integer) (end : Integer) (mode : (Matching_Mode.First | Matching_Mode.Last | Regex_Mode) = Regex_Mode.All)
+ Bounded (start : Integer) (end : Integer) (mode : (Matching_Mode | Regex_Mode) = Regex_Mode.All)

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Regex_Mode.enso:7]: Once the `First` and `Last` are not standalone types, technically `Matching_Mode.First` cannot show up in the signature. It probably won't be verified yet, as our typesystem is practically nonexistent at this moment, but I think it would be good to at least try to keep ourselves up-to-date with the design (which currently disallows just a single variant of an ADT being used in type signatures).

So if you could please find usages of `Matching_Mode.First | Matching_Mode.Last` in type signatures and replace them with just `Matching_Mode`.

Also if anywhere we had just one of these variants - now this kind of assumption is inexpressible in the type signature (I think this was the reason why we had them be separate standalone types). If we have functions that assume just one variant, we probably need to:
1. Update the signature to be valid, i.e. take the "whole" `Matching_Mode`
2. Ideally, make sure that the code can handle the earlier unsupported matching mode gracefully - for example instead of inexhaustive pattern match, I'd suggest to throw an `Illegal_Argument.Error` explaining that the other mode is not supported.
3. Would be good to ensure the documentation makes it clear which modes are supported and which are not - since the type signature no longer provides this information.

I'm not sure if there even are instances where this is an issue, but please do check.

File: test/Tests/src/Data/Text_Spec.enso [hunk 4]
```diff
@@ -1116,90 +1124,89 @@ spec =
...
+            abc.match "" matcher=regex mode=Matching_Mode.Last . should_equal ""
 
         Test.specify "should handle overlapping matches as shown in the examples" <|
-            "aaa".locate "aa" mode=Matching_Mode.Last matcher=Text_Matcher.Case_Sensitive . should_equal (Span.Value (1.up_to 3) "aaa")
-            "aaa".locate "aa" mode=Matching_Mode.Last matcher=Regex_Matcher.Value . should_equal (Span.Value (0.up_to 2) "aaa")
+            "aaa".locate "aa" mode=Matching_Mode.Last case_sensitivity=Case_Sensitivity.Sensitive . should_equal (Span.Value (1.up_to 3) "aaa")
 
-            "aaa aaa".locate "aa" mode=Matching_Mode.Last matcher=Text_Matcher.Case_Sensitive . should_equal (Span.Value (5.up_to 7) "aaa aaa")
-            "aaa aaa".locate "aa" mode=Matching_Mode.Last matcher=Regex_Matcher.Value . should_equal (Span.Value (4.up_to 6) "aaa aaa")
+            "aaa aaa".locate "aa" mode=Matching_Mode.Last case_sensitivity=Case_Sensitivity.Sensitive . should_equal (Span.Value (5.up_to 7) "aaa aaa")
```
Comment (radeusgd) [inline test/Tests/src/Data/Text_Spec.enso:1209]: Please update these as in examples - we can retain these `locate` tests but let's also add the new `match` tests based on `aAa`.

**Review by radeusgd — APPROVED** [review]
Looks good.

I'd just update the examples for `match` because currently they do not really illustrate any difference.

And I'd appreciate checking if references to `Matching_Mode` in type signatures are correctly updated after the change of this type.

File: test/Tests/src/Data/Text_Spec.enso (hunk 4 above)
Comment (GregoryTravis) [reply test/Tests/src/Data/Text_Spec.enso:1209]: Done

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Regex_Mode.enso (hunk 3 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Regex_Mode.enso:7]: It turns out there aren't any instances of this, but there was one false one, an unimplemented interface method, which probably should have had both First and Last in its signature. The single implementation of this interface method that was changed in this PR had both First and Last.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Pattern.enso [hunk 5]
```diff
@@ -69,7 +69,7 @@ type Pattern
 
        This method will _always_ return a vector. If no splits take place, the
        vector will contain a single element.
-    split : Text -> (Matching_Mode.First | Integer | Regex_Mode) -> Vector Text
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Pattern.enso:72]: @radeusgd This was the only case I found that had First but not Last. Since it's meant to be unimplemented, I won't check that the parameter is First. (And I suspect that Last should have been included here anyway.)

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso [hunk 6]
```diff
@@ -465,7 +465,7 @@ type Pattern
                  pattern = engine.compile "a" []
                  input = "bacadaeaf"
                  pattern.match input
-    split : Text -> Matching_Mode.First | Matching_Mode.Last | Regex_Mode | Integer -> Vector Text
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso:468]: @radeusgd This is the only implementation of the interface method that was changed in this PR, and it already had both First and Last.

File: test/Tests/src/Data/Text_Spec.enso [hunk 7]
```diff
@@ -1116,97 +1124,99 @@ spec =
...
 
         Test.specify "should allow to match one or more occurrences of a pattern in the text" <|
             "abacadae".match_all "a[bc]" . should_equal ["ab", "ac"]
             "abacadae".match_all "a." . should_equal ["ab", "ac", "ad", "ae"]
             "abacadae".match_all "a.*" . should_equal ["abacadae"]
             "abacadae".match_all "a.+?" . should_equal ["ab", "ac", "ad", "ae"]
 
+            "aAa".match "aa" mode=Matching_Mode.Last matcher=Text_Matcher.Case_Insensitive . should_equal "Aa"
+            "aAa".match "aa" mode=Matching_Mode.Last matcher=(Regex_Matcher.Value case_sensitivity=Case_Sensitivity.Insensitive) . should_equal "aA"
```
Comment (GregoryTravis) [inline test/Tests/src/Data/Text_Spec.enso:1218]: @radeusgd I would expect this one to return "Aa"? Is this the difference between Text and Regexp behavor for Matching_Mode.Last that's mentioned in the comments?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Pattern.enso (hunk 5 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Pattern.enso:72]: Okay, thanks for an exhaustive check. I suspect we have updated the implementation but forgot to update the 'interface' and since the latter does not really have a code connection to the former, its more of a documentation, we did not notice.

File: test/Tests/src/Data/Text_Spec.enso (hunk 7 above)
Comment (radeusgd) [reply test/Tests/src/Data/Text_Spec.enso:1218]: That's the whole point of this example and the comment.

What is the _actual_ behaviour here? Is the test passing in the current form or not? I'd expect it to be passing in _this_ form, unless something has changed.

```
   ! Last Match in Regex Mode
     Regex always performs the search from the front and matching the last
     occurrence means selecting the last of the matches while still generating
     matches from the beginning. This will lead to slightly different behavior
     for overlapping occurrences of a pattern in Regex mode than in exact text
     matching mode where the matches are searched for from the back.
```

This is explained in the comment that you moved around as part of this. However, maybe the explanation is not that good - if you have ideas how we could rephrase it to make it more understandable, I will appreciate that.

The thing is that Regex does not have a mode for 'matching backwards', all we can do is match all and pick the last one. And also Regex does not return overlapping matched - it will return a match at some position and then continue the search _after_ that match.

I.e. if I do `"abcde".match_all` I will get `["ab", "cd"]` and not for example `["ab", "bc", "cd", "de"]`. So in our case matching `aAa` against `aa` case insensitively has only one match `aA` which is returned.

This is what this test verifies and this is what the doc comment together with that example above tried to explain. Maybe we can improve the explanation to make it more understandable to our users, but I'm not sure how to phrase this stuff better.

Comment (GregoryTravis) [reply test/Tests/src/Data/Text_Spec.enso:1218]: Yes, that is what I wasn't sure about -- I've added your sentence to the comment: "And also Regex does not return overlapping matched - it will return a match at some position and then continue the search after that match."
